### PR TITLE
Deprecate passing a class to CapabilityManager::register

### PR DESCRIPTION
**Pull request by josephcsible**

### CODE REVIEW

**Comment by liach** [comment]
I'd suggest changing that callable to supplier in 1.13 since 1.13's change is huge.

File: src/main/java/net/minecraftforge/common/capabilities/CapabilityManager.java [hunk 1]
```diff
@@ -49,7 +49,10 @@
      * @param type The Interface to be registered
      * @param storage A default implementation of the storage handler.
      * @param implementation A default implementation of the interface.
+     * @deprecated Pass a constructor reference instead of a class name as the third parameter
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/capabilities/CapabilityManager.java]: I'd just mention to use the `register` method that takes an explicit factory. There is no requirement to use a constructor reference, although that is a nice way to do it.

**Review by diesieben07 — APPROVED** [review]

**Comment by LexManos** [comment]
Why? The point is to have both available if people feel like using one over the other. There really isn't a reason to remove existing code.

**Comment by josephcsible** [comment]
There's no real reason for it since anywhere .class would work, so would ::new, and there's also a slight performance difference, since we need to do reflection the .class way.

**Comment by LexManos** [comment]
Except for in-code, removing of this function {yes I know its just deprecated for now} would break existing mods.
As for the performance, that might be a concern if things were being abused a lot. But honestly I don't see that as the case. As the implementation is mainly a default reference meant more for documentation then for functionality and in my experience  the bulk of people provide custom implementations.

**Comment by josephcsible** [comment]
Isn't 1.13 going to break everything anyway? If you don't like that either, how about just deprecating it but not marking it for removal?

**Comment by liach** [comment]
@josephcsible You should add `[1.13]` in the title to indicate that this pull is intended for 1.13

**Comment by josephcsible** [comment]
It isn't intended for 1.13. I want to deprecate it in 1.12 (which this does), then maybe remove it in 1.13.

**Comment by mezz** [comment]
`::new` is better because it doesn't require the implicit contract of "you must have a no-args constructor" so it's a lot harder to mess up. Also when you construct from .class, the IDE thinks the no-args constructor is unused and it's easy to forget why you had it and remove it. 

I'm not super concerned about this particular case, but I've run into issues with the .class `newInstance()` constructor stuff in other places.

**Pull request merged**